system_tests/common_test.go

```diff
@@ -60,6 +60,182 @@ import (
+	return TransferBalanceTo(t, from, to, amount, tc.Info, tc.Client, tc.ctx)
+}
+
+func (tc *TestClient) GetBalance(t *testing.T, account common.Address) *big.Int {
+	return GetBalance(t, tc.ctx, tc.Client, account)
+}
+
+func (tc *TestClient) GetBaseFeeAt(t *testing.T, blockNum *big.Int) *big.Int {
+	return GetBaseFeeAt(t, tc.Client, tc.ctx, blockNum)
+}
+
+func (tc *TestClient) SendWaitTestTransactions(t *testing.T, txs []*types.Transaction) {
+	SendWaitTestTransactions(t, tc.ctx, tc.Client, txs)
+}
+
+func (tc *TestClient) DeploySimple(t *testing.T, auth bind.TransactOpts) (common.Address, *mocksgen.Simple) {
+	return deploySimple(t, tc.ctx, auth, tc.Client)
+}
+
+type NodeBuilder struct {
+	// NodeBuilder configuration
+	ctx           context.Context
+	Info          info
+	chainConfig   *params.ChainConfig
+	nodeConfig    *arbnode.Config
+	stackConfig   *node.Config
+	cachingConfig *execution.CachingConfig
+
+	// L1, L2 Node parameters
+	isSequencer   bool
+	takeOwnership bool
+	withL1        bool
+
+	// Blockchain parameters
+	dataDir     string
+	initMessage *arbostypes.ParsedInitMessage
+
+	// Created nodes
+	L1 *TestClient
+	L2 *TestClient
+}
+
+func NewNodeBuilder(ctx context.Context) *NodeBuilder {
+	return &NodeBuilder{ctx: ctx}
+}
+
+func (b *NodeBuilder) DefaultConfig(withL1 bool, n *arbnode.Config, c *params.ChainConfig) *NodeBuilder {
+	// most used values across current tests are set here as default
+	b.withL1 = withL1
+	if withL1 {
+		b.isSequencer = true
+		b.nodeConfig = arbnode.ConfigDefaultL1Test()
+		if n != nil {
+			b.nodeConfig = n
+		}
+		b.chainConfig = c
+	} else {
+		b.takeOwnership = true
+		b.nodeConfig = arbnode.ConfigDefaultL2Test()
+		if n != nil {
+			b.nodeConfig = n
+		}
+		b.chainConfig = c
+	}
+	return b
+}
+
+func (b *NodeBuilder) Build(t *testing.T) *NodeBuilder {
+	if b.withL1 {
+		l1, l2 := NewTestClient(b.ctx), NewTestClient(b.ctx)
+		l2.Info, l2.Node, l2.Client, l2.Stack, l1.Info, l1.Backend, l1.Client, l1.Stack =
+			createTestNodeOnL1WithConfigImpl(t, b.ctx, b.isSequencer, b.nodeConfig, b.chainConfig, b.stackConfig, b.Info)
+		b.L1, b.L2 = l1, l2
+	} else {
+		l2 := NewTestClient(b.ctx)
+		l2.Info, l2.Node, l2.Client =
+			CreateTestL2WithConfig(t, b.ctx, b.Info, b.nodeConfig, b.takeOwnership)
+		b.L2 = l2
+	}
+	return b
+}
+
+type SecondNodeParams map[string]interface{}
+
+func (b *NodeBuilder) Build2ndNode(t *testing.T, params SecondNodeParams) *TestClient {
+	if b.L1 == nil {
+		t.Fatal("builder did not previously build a L1 Node")
+	}
+	if b.L2 == nil {
+		t.Fatal("builder did not previously build a L2 Node")
+	}
+	if _, ok := params["dasConfig"]; ok {
+		nodeConf := arbnode.ConfigDefaultL1NonSequencerTest()
+		if params["dasConfig"] == nil {
+			nodeConf.DataAvailability.Enable = false
+		} else {
+			nodeConf.DataAvailability = *params["dasConfig"].(*das.DataAvailabilityConfig)
+		}
+		params["nodeConfig"] = nodeConf
+	} else if _, ok := params["nodeConfig"]; !ok {
```

Comment:
I'd start by checking nodeconfig (same logic as now), then check dasnodeconfig - so if dasnodeconfig is set - we start with nodeconfig that was set in previous step (params or builder's nodeconfig).